# Practicumopdrachten Week 2.3 (eerste kans)

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige practicum de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [49]:
import warnings
warnings.simplefilter('ignore')
import sqlite3
import pandas as pd
import numpy as np

Zet het bestand go_sales.sqlite in hetzelfde mapje als dit notebook.

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige practicum de code om een connectie met het bestand go_sales.sqlite te maken.

In [50]:
database_connection = sqlite3.connect('go_sales.sqlite')

Lees van de ingelezen go_sales-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- target
- returned_item
- return_reason

In [51]:
product = pd.read_sql_query("SELECT * FROM product", database_connection)
product_type = pd.read_sql_query("SELECT * FROM product_type", database_connection)
product_line = pd.read_sql_query("SELECT * FROM product_line", database_connection)
sales_staff = pd.read_sql_query("SELECT * FROM sales_staff", database_connection)
sales_branch = pd.read_sql_query("SELECT * FROM sales_branch", database_connection)
retailer_site = pd.read_sql_query("SELECT * FROM retailer_site", database_connection)
country = pd.read_sql_query("SELECT * FROM country", database_connection)
order_header = pd.read_sql_query("SELECT * FROM order_header", database_connection)
order_details = pd.read_sql_query("SELECT * FROM order_details", database_connection)
sales_targetdata = pd.read_sql_query("SELECT * FROM sales_targetdata", database_connection)
returned_item = pd.read_sql_query("SELECT * FROM returned_item", database_connection)
return_reason = pd.read_sql_query("SELECT * FROM return_reason", database_connection)

Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [22]:
product['PRODUCTION_COST'] = pd.to_numeric(product['PRODUCTION_COST'])
product_kosten_tussen_100_en_50 = (product['PRODUCTION_COST'] < 100) & (product['PRODUCTION_COST'] > 50)
product.loc[product_kosten_tussen_100_en_50, ['PRODUCT_NUMBER', 'PRODUCTION_COST']]     

PRODUCT_NUMBER  PRODUCTION_COST
13            110            60.00
14            111            81.80
17            114            80.00
24             17            60.00
25             18            86.00
28             20            78.55
31             23            65.33
32             24            62.50
56             46            52.86
58             48            55.00
69             58            57.13
71              6            85.11
81             69            73.33
83             70            80.00
84             71            60.00
85             72            72.50
88             75            80.00
91             78            79.19
92             79            92.58
93              8            75.00
94             80            94.12
97             83            56.25
98             84            78.55

Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [23]:
product['MARGIN'] = pd.to_numeric(product['MARGIN'])
product_margin_buiten_20_en_60 = (product['MARGIN'] < 0.20) | (product['MARGIN'] > 0.60)
product.loc[product_margin_buiten_20_en_60, ['PRODUCT_NUMBER', 'MARGIN']]    

PRODUCT_NUMBER  MARGIN
29              21    0.17
32              24    0.17
100             86    0.67
101             87    0.67
102             88    0.70
103             89    0.63
105             90    0.63

Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [24]:
country_currency_name_francs = country['CURRENCY_NAME'] == 'francs'
country.loc[country_currency_name_francs, ['COUNTRY']].sort_values(['COUNTRY'])

COUNTRY
7       Belgium
0        France
19  Switzerland

Geef een overzicht met daarin de verschillende introductiedatums waarop producten met meer dan 50% marge worden geïntroduceerd (1 kolom, 7 rijen) 

In [25]:
product['MARGIN'] = pd.to_numeric(product['MARGIN'])
product_introductiedatums_margin_meer_50 = (product['MARGIN'] > 0.50)
product.loc[product_introductiedatums_margin_meer_50, ['INTRODUCTION_DATE']].drop_duplicates(['INTRODUCTION_DATE'])

INTRODUCTION_DATE
4          10-12-2019
6          18-12-2019
7          27-12-2019
9           13-1-2020
14         15-12-2019
79         26-10-2016
100         15-2-2011

Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 28 rijen)

In [26]:
sales_branch_adres_regio = (sales_branch['ADDRESS2'].notna()) & (sales_branch['REGION'].notna())
sales_branch.loc[sales_branch_adres_regio, ['ADDRESS1', 'CITY']]

ADDRESS1              CITY
10  Prol. Paseo de la Reforma No. 51  Distrito Federal
11               202-2-3 Hyakunincho             Tokyo
12                     543-225 Asahi        Osaka City
13                  2315 Queen's Ave         Melbourne
15             Avenida Paulista, 333         São Paulo
21                    3 Albany Court        Birmingham
27               Singelgravenplein 4         Amsterdam

Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

In [27]:
country_dollars = (country['CURRENCY_NAME'] == 'dollars') | (country['CURRENCY_NAME'] == 'new dollar')
country.loc[country_dollars, ['COUNTRY']].sort_values(['COUNTRY'])

COUNTRY
6       Australia
15         Canada
3          Taiwan
14  United States

Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen) 

In [28]:
retailer_site_adressen_stad_postcode_D = (retailer_site['POSTAL_ZONE'].str[0] == 'D') & (retailer_site['ADDRESS2'].notna())
retailer_site.loc[retailer_site_adressen_stad_postcode_D, ['ADDRESS1', 'ADDRESS2', 'CITY']].sort_values(['ADDRESS2'])

ADDRESS1       ADDRESS2       CITY
205  Röntgenstraße 90  3. Tür rechts  Frankfurt
351   Grubesallee 141       4. Stock    Hamburg

## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 

In [59]:
returned_item["RETURN_QUANTITY"] = returned_item["RETURN_QUANTITY"].astype(int)
returned_item = returned_item["RETURN_QUANTITY"].sum()
pd.DataFrame({'TOTAAL': [returned_item]})

TOTAAL
0   14306

Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

In [60]:
retailer_site = retailer_site['REGION'].count()
pd.DataFrame({'REGIOS': [retailer_site]})

REGIOS
0     251

Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

In [5]:
product['MARGIN'] = pd.to_numeric(product['MARGIN'], errors='coerce')

laagste_marge = product['MARGIN'].min()
hoogste_marge = product['MARGIN'].max()
gemiddelde_marge = product['MARGIN'].mean().round(2)
pd.DataFrame({'LAAGSTE_MARGE': [laagste_marge], 'HOOGSTE_MARGE': [hoogste_marge],'GEMIDDELDE_MARGE': [gemiddelde_marge]})


LAAGSTE_MARGE  HOOGSTE_MARGE  GEMIDDELDE_MARGE
0           0.17            0.7               0.4

Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

In [4]:
retailer_site['CITY'] = (retailer_site['ADDRESS2'] == 'NULL')
retailer_site = retailer_site['CITY'].count()
pd.DataFrame({'AANTAL': [retailer_site]})

TOTAAL
0     391

Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

In [11]:
order_details['UNIT_COST'] = (order_details['UNIT_SALE_PRICE']) < (order_details['UNIT_PRICE'])
order_details = order_details['UNIT_COST'].mean()
pd.DataFrame({'GEMIDDELD': [order_details]})

GEMIDDELD
0   0.325755

Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

In [36]:
aantal_per_functie = sales_staff.groupby('POSITION_EN', as_index= False)['SALES_STAFF_CODE'].count()
aantal_per_functie.columns = ['SALES_STAFF_CODE', 'POSITION_EN']
aantal_per_functie

SALES_STAFF_CODE  POSITION_EN
0                Branch Manager           19
1              District Manager            4
2               General Manager            1
3  Level 1 Sales Representative           13
4  Level 2 Sales Representative           39
5  Level 3 Sales Representative           24
6              Regional Manager            2

Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

In [28]:
aantal_per_telefoonnummer = sales_staff.groupby('WORK_PHONE', as_index= False)['SALES_STAFF_CODE'].count()
aantal_per_telefoonnummer = aantal_per_telefoonnummer[aantal_per_telefoonnummer['SALES_STAFF_CODE'] > 4]
aantal_per_telefoonnummer

WORK_PHONE  SALES_STAFF_CODE
4     +(44) 121 3505267                 5
10  +(61) 03 2982  4242                 5
11  +31 (0)20 692 93 94                 5
13    +33 1 68 94 52 20                 5
18     1 (206) 292-0012                 5
19     1 (305) 557-4810                 6
20     1 (310) 281-5722                 5
21     1 (403) 232-5986                 5
22     1 (416) 493-5595                 5
23     1 (617) 268-6754                 6

## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

In [36]:
merged_retailer_site_and_country = pd.merge(retailer_site, country)
merged_retailer_site_and_country = merged_retailer_site_and_country[merged_retailer_site_and_country['COUNTRY'] == 'Netherlands']
merged_retailer_site_and_country.groupby('ADDRESS1', as_index=False)['CITY'].first()

ADDRESS1   CITY
0      Amsterdamsestraatweg 8  False
1                    Beets 36  False
2                Duivendaal 1  False
3             Gooiergracht 91  False
4               Havenbuurt 12  False
5               Hogehilweg 13  False
6             Hoofdstraat 150  False
7              Kalfjeslaan 47  False
8         Krayenhoffstraat 11  False
9   Laan van Meerdervoort 966  False
10      Noorderspoorsingel 28  False
11                Olde Diek 2  False
12         Overschiestraat 19  False
13               Perestraat 1  False
14             Spoorstraat 43  False
15         Stadhouderslaan 18  False
16               Startbaan 11  False
17              Voorstraat 37  False
18             Westhavenweg 8  False
19              Westzeedijk 4  False

Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen) 

In [37]:
merged_product_and_product_type = pd.merge(product, product_type)
merged_product_and_product_type = merged_product_and_product_type[merged_product_and_product_type['PRODUCT_TYPE_EN'] == 'Eyewear'] \
    [['PRODUCT_NAME']]
merged_product_and_product_type

PRODUCT_NAME
80      Polar Sun
81      Polar Ice
83   Polar Sports
84     Polar Wave
85  Polar Extreme

Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 

In [84]:
merged_retailer_site_and_sales_staff = pd.merge(retailer_site, sales_staff)
merged_retailer_site_and_sales_staff = merged_retailer_site_and_sales_staff[merged_retailer_site_and_sales_staff['POSITION_EN'] == 'Branch Manager'] 
merged_retailer_site_and_sales_staff[['ADDRESS1', 'FIRST_NAME', 'LAST_NAME']].drop_duplicates(['ADDRESS1']).drop_duplicates(['FIRST_NAME']).drop_duplicates(['LAST_NAME'])

ADDRESS1 FIRST_NAME LAST_NAME
7  1117 Franklin Blvd     Walter    Taylor

Geef een overzicht met daarin van de verkopers hun functie en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 9 rijen) 

In [30]:
merged_sales_staff_and_order_header = pd.merge(sales_staff, order_header, left_on = 'SALES_STAFF_CODE', how = 'left', right_on = 'SALES_STAFF_CODE')
 
sales_staff_manager_order_date = merged_sales_staff_and_order_header['POSITION_EN'].str.contains('Manager')
merged_sales_staff_and_order_header.loc[sales_staff_manager_order_date, ['POSITION_EN', 'ORDER_DATE']].drop_duplicates(['POSITION_EN','ORDER_DATE'])

POSITION_EN  ORDER_DATE
421     Branch Manager         NaN
422    General Manager         NaN
1276  Regional Manager         NaN
1643  District Manager         NaN
5325    Branch Manager  2022-09-09
5326    Branch Manager  2022-10-31
5327    Branch Manager  2022-12-13
5328    Branch Manager  2022-03-13
5329    Branch Manager  2022-01-09

Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

In [33]:
merged_product_product_type_and_order_details = pd.merge(product, product_type).merge(order_details)
merged_product_product_type_and_order_details = merged_product_product_type_and_order_details[merged_product_product_type_and_order_details['QUANTITY'] > 750]
merged_product_product_type_and_order_details[['PRODUCT_NAME', 'PRODUCT_TYPE_EN']].drop_duplicates(['PRODUCT_NAME'])

PRODUCT_NAME    PRODUCT_TYPE_EN
8616                Star Peg              Tents
15975              Firefly 2           Lanterns
37903        BugShield Spray  Insect Repellents
38313  BugShield Lotion Lite  Insect Repellents
38770       BugShield Lotion  Insect Repellents
39618      BugShield Extreme  Insect Repellents
40754         Sun Shelter 15          Sunscreen
41106         Sun Shelter 30          Sunscreen
41681             Sun Shield          Sunscreen

Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

In [77]:
merged_product_and_order_details = pd.merge(product, order_details)
merged_product_and_order_details['UNIT_PRICE'] = pd.to_numeric(merged_product_and_order_details['UNIT_PRICE'], errors='coerce')
merged_product_and_order_details['UNIT_SALE_PRICE'] = pd.to_numeric(merged_product_and_order_details['UNIT_SALE_PRICE'], errors='coerce')

product_has_more_then_40_precent_discount = ((merged_product_and_order_details['UNIT_PRICE'] - merged_product_and_order_details['UNIT_SALE_PRICE']) / merged_product_and_order_details['UNIT_PRICE']) > 0.4
merged_product_and_order_details.loc[product_has_more_then_40_precent_discount, ['PRODUCT_NAME']].drop_duplicates('PRODUCT_NAME')

PRODUCT_NAME
2765   Hailstorm Titanium Woods Set
37446             BugShield Natural
37967               BugShield Spray
38306         BugShield Lotion Lite
38772              BugShield Lotion
39561             BugShield Extreme
40799                Sun Shelter 15
41194                Sun Shelter 30

Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

In [80]:
merged_return_reason_and_returned_item = pd.merge(return_reason, returned_item).merge(order_details)
merged_return_reason_and_returned_item['RETURN_QUANTITY'] = pd.to_numeric(merged_return_reason_and_returned_item['RETURN_QUANTITY'], errors='coerce')
merged_return_reason_and_returned_item['QUANTITY'] = pd.to_numeric(merged_return_reason_and_returned_item['QUANTITY'], errors='coerce')

product_has_more_then_90_precent_discount = (merged_return_reason_and_returned_item['RETURN_QUANTITY'] / merged_return_reason_and_returned_item['QUANTITY']) > 0.9
merged_return_reason_and_returned_item.loc[product_has_more_then_90_precent_discount, ['RETURN_DESCRIPTION_EN']].drop_duplicates('RETURN_DESCRIPTION_EN')

RETURN_DESCRIPTION_EN
451   Wrong product ordered
521   Wrong product shipped
583  Unsatisfactory product

## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

In [17]:
merged_product_and_product_type = pd.merge(product, product_type)
merged_product_and_product_type.groupby('PRODUCT_TYPE_EN', as_index=False)['PRODUCT_NUMBER'].count()

PRODUCT_TYPE_EN  PRODUCT_NUMBER
0             Binoculars               4
1   Climbing Accessories               7
2           Cooking Gear              10
3                Eyewear               5
4              First Aid               5
5       Golf Accessories               4
6      Insect Repellents               5
7                  Irons               4
8                 Knives               5
9               Lanterns              12
10            Navigation               4
11                 Packs               6
12               Putters               3
13                  Rope               4
14                Safety               4
15         Sleeping Bags               7
16             Sunscreen               5
17                 Tents               6
18                 Tools               6
19               Watches               5
20                 Woods               4

Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

In [68]:
merged_country_and_retailer_site = pd.merge(country, retailer_site)
merged_country_and_retailer_site.groupby('COUNTRY', as_index=False)['CITY'].count()

COUNTRY  CITY
0        Australia    10
1          Austria    10
2          Belgium    10
3           Brazil     1
4           Canada    31
5            China    10
6          Denmark     6
7          Finland     5
8           France    29
9          Germany    32
10           Italy    12
11           Japan    25
12           Korea    10
13          Mexico    10
14     Netherlands    20
15           Spain     9
16          Sweden    12
17     Switzerland    13
18          Taiwan    10
19  United Kingdom    25
20   United States   101

Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (4 kolommen, 10 rijen) 

In [69]:
merged_product_product_type_and_order_details = pd.merge(product, product_type).merge(order_details)
merged_product_product_type_and_order_details['UNIT_PRICE'] = pd.to_numeric(merged_product_product_type_and_order_details['UNIT_PRICE'], errors='coerce')

merged_product_product_type_and_order_details = merged_product_product_type_and_order_details[merged_product_product_type_and_order_details['PRODUCT_TYPE_EN'] == 'Cooking Gear'] 
merged_product_product_type_and_order_details = merged_product_product_type_and_order_details.groupby(['PRODUCT_NAME', 'PRODUCT_TYPE_EN'], as_index=False).agg({'QUANTITY': 'sum', 'UNIT_PRICE': 'mean'})
merged_product_product_type_and_order_details 

PRODUCT_NAME PRODUCT_TYPE_EN  QUANTITY  UNIT_PRICE
0          TrailChef Canteen    Cooking Gear     31712   12.178329
1         TrailChef Cook Set    Cooking Gear     12700   52.605173
2              TrailChef Cup    Cooking Gear     31834    7.137756
3  TrailChef Deluxe Cook Set    Cooking Gear      3744  123.537733
4     TrailChef Double Flame    Cooking Gear     14510  136.601562
5           TrailChef Kettle    Cooking Gear      9900   12.375391
6      TrailChef Kitchen Kit    Cooking Gear     25964   22.744308
7     TrailChef Single Flame    Cooking Gear     10614   65.090826
8         TrailChef Utensils    Cooking Gear     15928   17.731438
9        TrailChef Water Bag    Cooking Gear     41770    6.181641

Geef een overzicht met daarin van de verkopers behorend tot de verkoopafdeling uit ‘Amsterdam’ de voornaam en achternaam van de verkoper en zijn laagste en hoogste verkooptarget van 2006. (4 kolommen, 4 rijen) 

In [70]:
merged_sales_staff_sales_branch_and_sales_targetdata = pd.merge(sales_branch, sales_staff).merge(sales_targetdata)

merged_sales_staff_sales_branch_and_sales_targetdata = merged_sales_staff_sales_branch_and_sales_targetdata[(merged_sales_staff_sales_branch_and_sales_targetdata['CITY'] == 'Amsterdam') & (merged_sales_staff_sales_branch_and_sales_targetdata['SALES_YEAR'] == 2006)]
merged_sales_staff_sales_branch_and_sales_targetdata = merged_sales_staff_sales_branch_and_sales_targetdata.groupby(['FIRST_NAME', 'LAST_NAME'], as_index=False)['SALES_TARGET'].agg(['min', 'max'])
merged_sales_staff_sales_branch_and_sales_targetdata  

#werkt wel met correcte output:
#merged_sales_staff_sales_branch_and_sales_targetdata = pd.merge(sales_branch, sales_staff).merge(sales_targetdata)

#merged_sales_staff_sales_branch_and_sales_targetdata = merged_sales_staff_sales_branch_and_sales_targetdata[merged_sales_staff_sales_branch_and_sales_targetdata['CITY'] == 'Amsterdam']
#merged_sales_staff_sales_branch_and_sales_targetdata = merged_sales_staff_sales_branch_and_sales_targetdata.groupby(['FIRST_NAME', 'LAST_NAME'], as_index=False)['SALES_TARGET'].agg(['min', 'max'])
#merged_sales_staff_sales_branch_and_sales_targetdata  

Empty DataFrame
Columns: [FIRST_NAME, LAST_NAME, min, max]
Index: []

Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

In [29]:
merged_country_retailer_site = pd.merge(country, sales_branch).merge(sales_staff)

merged_country_retailer_site = merged_country_retailer_site.groupby(['COUNTRY', 'CITY'], as_index=False)['CITY'].agg(['count'])
merged_country_retailer_site = merged_country_retailer_site.rename(columns={'CITY': 'verkoper', 'count': 'klanten'})

merged_country_retailer_site

COUNTRY          verkoper  klanten
0        Australia         Melbourne        5
1          Austria              Wien        2
2          Belgium          Heverlee        4
3           Brazil         São Paulo        3
4           Canada           Calgary        5
5           Canada           Toronto        5
6            China          Shanghai        2
7          Finland            Kuopio        2
8           France              Lyon        4
9           France             Paris        4
10         Germany           Hamburg        4
11         Germany           München        3
12           Italy            Milano        3
13           Japan        Osaka City        2
14           Japan             Tokyo        2
15           Korea             Seoul        2
16          Mexico  Distrito Federal        3
17     Netherlands         Amsterdam        5
18           Spain            Bilbao        2
19          Sweden             Kista        4
20     Switzerland            Zürich        3
21          Taiwan         San Chung        2
22  United Kingdom        Birmingham        5
23  United Kingdom            London        4
24   United States            Boston        6
25   United States       Los Angeles        6
26   United States             Miami        5
27   United States           Seattle        5

## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

In [141]:
sales_staff_nooit_verkocht = (sales_staff['EXTENSION'].notna()) & (sales_staff['POSITION_EN'].str.contains('Manager'))
sales_staff.loc[sales_staff_nooit_verkocht, ['FIRST_NAME', 'LAST_NAME']]

FIRST_NAME      LAST_NAME
7      Walter         Taylor
8    Penelope        Tamrine
11      Polly         Linton
19      Gracy        Gellens
26       Else         Mörike
27      Frank          Fuchs
30      Fritz         Hirsch
31       Jörg          Kunze
38       Kick        Kalkman
43      Karin      Bergström
44      Denis           Pagé
47      Sally          White
49      Frank        Bretton
50    Georges  Saint-Germain
56       Alex      Rodriguez
60   Samantha          Floyd
65      Julie          Olsen
69      Corey         Wright
76    Estelle        Johnson
77       Greg        Belding
80      Julie        Béchard
84      Lucía          Reyna
89     Kazumi        Uragome
91      Alice         Walter
98     Bayard          Lopes

Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (1 kolom, 2 rijen) 

In [5]:
product['MARGIN'] = pd.to_numeric(product['MARGIN'], errors='coerce')

gemiddelde_marge = product['MARGIN'].mean()
marge_lager_dan_gemiddeld = product['MARGIN'] < gemiddelde_marge
geselecteerde_producten = product.loc[marge_lager_dan_gemiddeld, ['MARGIN']]
gemiddelde_marge_geselecteerde_producten = geselecteerde_producten['MARGIN'].mean()
pd.DataFrame({'MARGIN': [gemiddelde_marge, gemiddelde_marge_geselecteerde_producten]})

MARGIN
0  0.396609
1  0.292203

Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

In [55]:
order_details['UNIT_PRICE'] = pd.to_numeric(order_details['UNIT_PRICE'], errors='coerce')

merged_order_details_product = pd.merge(returned_item, order_details, left_on= 'ORDER_DETAIL_CODE', how='right', right_on= 'ORDER_DETAIL_CODE').merge(product, left_on= 'PRODUCT_NUMBER', how='left', right_on= 'PRODUCT_NUMBER')
producten_500_en_nooit_teruggebracht = (merged_order_details_product['UNIT_PRICE'] > 500) & (merged_order_details_product['RETURN_CODE'].isna())
merged_order_details_product.loc[producten_500_en_nooit_teruggebracht, ['PRODUCT_NAME']].drop_duplicates(['PRODUCT_NAME'])

PRODUCT_NAME
33               Hailstorm Steel Irons
36            Hailstorm Titanium Irons
41          Lady Hailstorm Steel Irons
44       Lady Hailstorm Titanium Irons
47        Hailstorm Titanium Woods Set
51           Hailstorm Steel Woods Set
55   Lady Hailstorm Titanium Woods Set
58      Lady Hailstorm Steel Woods Set
132                          Star Dome
136                       Star Gazer 2
138                       Star Gazer 3
141                       Star Gazer 6
263                     Husky Rope 200

Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

In [6]:
for index, row in sales_staff.iterrows():
    heeft_manager = row['POSITION_EN']
    
    if 'Manager' in heeft_manager:
        sales_staff.at[index, 'HEEFT_MANAGER'] = 'Ja'   
    else:
        sales_staff.at[index, 'HEEFT_MANAGER'] = 'Nee'  
         
sales_staff[['LAST_NAME', 'HEEFT_MANAGER']]

LAST_NAME HEEFT_MANAGER
0    Savolainen           Nee
1           Kim           Nee
2          Choi           Nee
3           Kao           Nee
4          Fang           Nee
..          ...           ...
97     Iglesias           Nee
98        Lopes            Ja
99       Barros           Nee
100   Guimarães           Nee
101    Lindholm           Nee

[102 rows x 2 columns]

Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [56]:
from datetime import date
date.today().year


2024

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [57]:
from datetime import datetime

date_str = '16-8-2013'
date_format = '%d-%m-%Y'
date_obj = datetime.strptime(date_str, date_format)

date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 25 jaar in dienst) of een ‘lang in dienst’ (groter gelijk dan 12 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 

In [59]:
sales_staff['DATE_HIRED'] = sales_staff['DATE_HIRED'].astype(str)

for index, row in sales_staff.iterrows():
    date_hired = row['DATE_HIRED']
    date_format = '%d-%m-%Y'
    date_obj = datetime.strptime(date_hired, date_format)

    years_in_service = (datetime.today().year - date_obj.year)
    
    if years_in_service >= 12:
        sales_staff.at[index, 'KORT_IN_DIENST'] = 'kort in dienst'
    else:
        sales_staff.at[index, 'LANG_IN_DIENST'] = 'lang in dienst'
        
sales_staff[['KORT_IN_DIENST', 'LANG_IN_DIENST']]    

KORT_IN_DIENST  LANG_IN_DIENST
0               NaN  lang in dienst
1               NaN  lang in dienst
2               NaN  lang in dienst
3               NaN  lang in dienst
4    kort in dienst             NaN
..              ...             ...
97              NaN  lang in dienst
98              NaN  lang in dienst
99              NaN  lang in dienst
100             NaN  lang in dienst
101             NaN  lang in dienst

[102 rows x 2 columns]